### DISCUSSION TITLE
Water97FluidProperty problem

### DISCUSSION THREAD

**Discussion by SomnusYu** [idaholab/moose#25651] [Q&A Modules: General] [answered]
Hi all,

I would like to calculate pressure using enthalpy and entropy (**p_from_h_s**)[h and s are known variables]. 
But Water97FluidProperty does not contain this function. What should I do? Should I use an iteration (assume p and T)?

Thanks!

**Comment by GiudGiud** [comment]
Hello

Yes you should try to implement it. 
An iterative method would work

Guillaume

**Reply by SomnusYu** [reply]
Do you have any idea how to implement the iteration method?

**Reply by GiudGiud** [reply, marked as answer]
Use the Newton1Dsolve method. See TabulatedFP for examples 

The idea is that as long as you have s(p,h) or h(p,s) defined, or if you can compute it in several steps, you can use that to compute the other one from a Newton Raphson solve

**Reply by SomnusYu** [reply]
Thank you very much! I find a function in SinglePhaseFluidProperties:
```
void
SinglePhaseFluidProperties::p_T_from_h_s(const T & h,     // h value
                                         const T & s,     // s value
                                         const Real p0,   // initial guess
                                         const Real T0,   // initial guess
                                         T & pressure,    // returned pressure
                                         T & temperature, // returned temperature
                                         bool & conversion_succeeded) const
```
I think I can use it directly to calculate pressure.

**Reply by GiudGiud** [reply]
this should be all you need indeed. It s iterative iirc so it might be a bit slow

The other alternative is to math it out, solve analytically for the property you need in the different regions of IAPWS IF 97

**Reply by SomnusYu** [reply]
However, I think the analytical expressions (p_from_h_s) in the different regions of IAPWS-IF97 are difficult for me to construct. Do you have the exact expressions, or where can I find the exact expressions? If yes, I could have a try.

**Reply by GiudGiud** [reply]
You can check the papers listed in the documentation.
https://mooseframework.inl.gov/source/fluidproperties/Water97FluidProperties.html
I think it will be difficult unless it's already written there fand you just need to implement

**Reply by SomnusYu** [reply]
Ok, thank you very much!

**Reply by cpgr** [reply]
There are backward equations for p(h,s) for some of the regions given by IAPWS - see http://www.iapws.org/release.html. These are much faster than iteration, but would require some hours of coding to add them. It isn't difficult, just time consuming, if you want to have a go at adding them.

**Reply by SomnusYu** [reply]
Thank you very much!
